src/storage/src/source/mysql/replication.rs

```diff
@@ -0,0 +1,385 @@
+    let replication_updates = data_stream
+        .as_collection()
+        .map(|(output_index, row)| (output_index, row.err_into()));
+
+    let errors = definite_errors.concat(&transient_errors.map(ReplicationError::from));
+
+    (
+        replication_updates,
+        upper_stream,
+        errors,
+        button.press_on_drop(),
+    )
+}
+
+/// Produces the logical replication stream while taking care of regularly sending standby
+/// keepalive messages with the provided `uppers` stream.
+///
+/// The returned stream will contain all transactions that whose commit LSN is beyond `resume_lsn`.
+async fn raw_stream<'a>(
+    config: &'a RawSourceCreationConfig,
+    mut conn: Conn,
+    // TODO: this is only needed while we use a single source-id for the timestamp; remove once we move to a partitioned timestamp
+    gtid_source_id: Uuid,
+    resume_transaction_id: TransactionId,
+) -> Result<Result<BinlogStream, DefiniteError>, TransientError> {
+    // Verify the MySQL system settings are correct for consistent row-based replication using GTIDs
+    // TODO: Should these return DefiniteError instead of TransientError?
+    ensure_gtid_consistency(&mut conn).await?;
+    ensure_full_row_binlog_format(&mut conn).await?;
+    ensure_replication_commit_order(&mut conn).await?;
+
+    // To start the stream we need to provide a GTID set of the transactions that we've 'seen'
+    // and the server will send us all transactions that have been committed after that point.
+    // NOTE: The 'Gno' intervals in this transaction-set use an open set [start, end)
+    // interval, which is different than the closed-set [start, end] form returned by the
+    // @gtid_executed system variable. So the intervals we construct in this GTID set
+    // end with the value of the transaction-id that we want to start replication at,
+    // which happens to be our `resume_transaction_id`
+    let seen_gtids = match resume_transaction_id.into() {
+        0 | 1 => {
+            // If we're starting from the beginning of the binlog or the first transaction id
+            // we haven't seen 'anything'
+            Vec::new()
+        }
+        ref a => {
+            // NOTE: Since we enforce replica_preserve_commit_order=ON we can start the interval at 1
+            // since we know that all transactions with a lower transaction id were monotonic
+            vec![Sid::new(*gtid_source_id.as_bytes()).with_interval(GnoInterval::new(1, *a))]
+        }
+    };
+
+    // Request that the stream provide us with a heartbeat message when no other messages have been sent
+    let heartbeat = Duration::from_secs(3);
+    conn.query_drop(format!(
+        "SET @master_heartbeat_period = {};",
+        heartbeat.as_nanos()
+    ))
+    .await?;
+
+    let repl_stream = match conn
+        .get_binlog_stream(
+            // TODO: The server-id should probably be something more static than the worker-id?
```

Comment:
Yes, for kafka client ids we use `materialize-{environment_id}-{connection_id}-{sink_id}` so here we could use `materialize-{environment_id}-{source_id}`

Reply:
Any reason not to include the connection ID, for consistency with Kafka?

Reply:
We could (and maybe we should). I thought the reason for including the connection id for Kafka sinks was that the connection included the progress topic. Do you remember why we included the connection id? It seems superfluous now that I look at it again since from a source/sink id there is one unambiguous connection object always

Reply:
It enables writing richer ACLs in Kafka! For example, you can write an ACL that targets all sources/sinks attached to a given connection, without knowing the specific IDs of those sources/sinks ahead of time.

Looking deeper though, it seems like the server ID needs to be an integer? So this whole discussion might be moot!

Reply:
Yep this needs to be a `u32`. I'm not convinced that this ID is very important in our setup, since it appears this `server-id` is used to break cycles in a replication topology if one exists when not using GTIDs. However the docs still say that we need this value to be 'different' than any other server-id in the replication topology (but it doesn't say what the consequence is if we did re-use one).

Reply:
Wait so does this need to be a user-facing `SERVER ID` parameter to `CREATE CONNECTION`?

Reply:
> Wait so does this need to be a user-facing SERVER ID parameter to CREATE CONNECTION?

Potentially. There is another confused person here https://dba.stackexchange.com/questions/325053/why-is-the-mysql-server-id-necessary-for-replication-when-the-server-uuid-exists with a fairly unsatisfying answer.

I'm going to see if I can just test this locally and see what happens if we reuse the server-id of the source server, or something similar

Reply:
When testing locally - using a server_id of `0` caused issues but using `1` (the same server-id as the mysql server) worked fine. So using a duplicate server_id doesn't actually cause issues when using GTID-based replication. Therefore I think we could just leave this at `1` or even set this to an arbitrary 'materialize' constant that we can document, which would aid customers in identifying which connection was ours when using the `SHOW REPLICAS` command: https://dev.mysql.com/doc/refman/8.0/en/show-replicas.html (as long as they didn't happen to pick the same arbitrary value for one of their replicas)

Reply:
I've pushed a change that uses the source-id plus an arbitrary offset (524000) so that we hopefully pick a unique value for the server-id. Then our customers can identify materialize connections by looking for connections with a server-id of 524000 + source-id. This is a best-effort approach, since if we do end up stepping on an existing server-id it doesn't appear that anything will go wrong (based on my testing). Let me know what y'all think.